WordPress/gutenberg
# Style engine: enqueue block supports styles in Gutenberg

## What?

A follow up to:
- https://github.com/WordPress/gutenberg/pull/42452

## What?
Let's enqueue stored block supports styles in Gutenberg, rather than in the style engine.

At the same time, we'll combine block style and layout styles into one style tag.

## Why?
This PR follows up on a [discussion](https://github.com/WordPress/gutenberg/pull/42452#discussion_r934132780) in #42452, the substance of which is that it's probably better to give the consumer control over how styles are rendered via hooks / enqueues etc

## How?
Adding a new function in `script-loader.php` - `gutenberg_enqueue_stored_block_supports_styles()` (name to be debated) that grabs block supports styles from the style engine store and enqueues the stylesheet.

## Testing Instructions

Test in both classic and block themes.

Check that elements and layouts styles are enqueued. 

Create a new post and insert some text links and group blocks (or any layout styles) 

Here's some test HTML:

<details>

<summary>Example</summary>

```html

<div class="wp-block-group has-background has-link-color" style="background-color:#ee6b6b">
<p class="has-link-color"><a href="https://word">Test</a></p>



<p class="has-link-color"><a href="https://word">Test</a></p>
</div>



<div class="wp-block-group has-background" style="background-color:#c4cce6">
<p class="has-link-color"><a href="https://word">Test</a></p>
</div>



<div class="wp-block-group has-background" style="background-color:#9eebbc">
<p class="has-link-color"><a href="https://word">Test</a></p>
</div>

```


</details>

<img width="902" alt="Screen Shot 2022-08-02 at 3 42 30 pm" src="https://user-images.githubusercontent.com/6458278/182300370-058392c3-8adf-4f51-abee-3d62b438845b.png">

Expected CSS output. For classic themes this style tag should appear at the bottom of the page. For block themes, in the head.

```html
<style id='block-supports-styles-inline-css'>
.wp-elements-bfe89bcbd3fbf76f9bf93ac884528d9a a {color: var(--wp--preset--color--vivid-cyan-blue);}.wp-elements-85694c8ee6aa1fb95efca209b615f0b0 a {color: var(--wp--preset--color--white);}.wp-elements-6ac1c4acebd6a433ed64dc4e6b0eb185 a {color: #eaf1f6;}.wp-elements-469734e762a60395c96370480f3c9414 a {color: var(--wp--preset--color--vivid-red);}.wp-elements-9307c4f13980ac2b051b4fcf5f7b28fa a {color: #cd31d6;}.wp-block-group.wp-container-1 {flex-wrap: nowrap; align-items: flex-start;}.wp-block-post-content.wp-container-4 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 840px; margin-left: auto !important; margin-right: auto !important;}.wp-block-post-content.wp-container-4 > .alignwide {max-width: 1100px;}.wp-block-group.wp-container-2 > :where(:not(.alignleft):not(.alignright):not(.alignfull)),.wp-block-group.wp-container-3 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 342px; margin-left: auto !important; margin-right: auto !important;}.wp-block-group.wp-container-2 > .alignwide,.wp-block-group.wp-container-3 > .alignwide {max-width: 342px;}.wp-block-group.wp-container-2 .alignfull,.wp-block-group.wp-container-3 .alignfull,.wp-block-post-content.wp-container-4 .alignfull {max-width: none;}
</style>
```
